src/hidra/receiver/asapo_transfer.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
+import logging
+
+from hidra import Transfer
+from plugins.asapo_producer import AsapoWorker
+
+logger = logging.getLogger(__name__)
+
+
+class AsapoTransfer:
+    def __init__(self, endpoint, beamtime, token, n_threads, file_regex,
+                 default_data_source=None, timeout=5, beamline='auto',
+                 start_file_idx=1):
+
+        self.signal_host = "localhost"
+        self.targets = [["localhost", "50101", 1]]
+        self.asapo_worker = AsapoWorker(endpoint, beamtime, token, n_threads, file_regex,
+                                        default_data_source, timeout, beamline,
+                                        start_file_idx)
+
+        self.start_transfer()
+
+    def start_transfer(self):
+
+        query = Transfer("STREAM_METADATA", self.signal_host)
+        query.initiate(self.targets)
+        query.start()
+        self.run(query)
+        query.stop()
+
+    def run(self, query):
+        while True:
+            try:
+                [metadata, data] = query.get()
+            except:
+                break
+
+            if metadata is not None and data is not None:
+                local_path = f'{metadata["relative_path"]}/{metadata["filename"]}'
```

Comment:
This should use `generate_filepath` from the `transfer` module

Reply:
Not `generate_filepath` is used. However I don't understand, what is `target_dir`.

Reply:
The `target_dir` will be a prefix to the complete path returned by `generate_filepath`. I think it is safe to set it to the empty string.